cpu/esp32/periph/i2c_hw.c

```diff
@@ -654,262 +530,124 @@ static void _i2c_read_cmd (i2c_t dev, uint8_t* data, uint8_t len, bool last)
+void _i2c_transfer_timeout(void *arg)
 {
     i2c_t dev = (i2c_t)(uintptr_t)arg;
 
-    /* reset the hardware if it I2C got stuck */
-    _i2c_reset_hw(dev);
+    /* reset the hardware if I2C got stuck */
+    i2c_hw_fsm_reset(dev);
 
     /* set result to timeout */
-    _i2c_bus[dev].results |= I2C_TIME_OUT_INT_ST;
+    _i2c_bus[dev].status = I2C_TIME_OUT_INT_ENA_M;
 
     /* wake up the thread that is waiting for the results */
-    thread_flags_set((thread_t*)thread_get(_i2c_bus[dev].pid), I2C_THREAD_FLAG);
+    mutex_unlock(&_i2c_bus[dev].cmd_lock);
 }
 
 /* Transfer of commands in I2C controller command pipeline */
-static void _i2c_transfer (i2c_t dev)
+static void _i2c_transfer(i2c_t dev)
 {
     DEBUG("%s cmd=%d\n", __func__, _i2c_bus[dev].cmd);
 
-    #if FIFO_USED
-    /* reset RX FIFO queue */
-    _i2c_hw[dev].regs->fifo_conf.rx_fifo_rst = 1;
-    /* cppcheck-suppress redundantAssignment
-     * Likely due to cppcheck not being able to located all headers, it misses
-     * the volatile qualifier. The assignments are to trigger a reset, but
-     * look like dead writes to tools unaware of volatile */
-    _i2c_hw[dev].regs->fifo_conf.rx_fifo_rst = 0;
-    #endif
-
     /* disable and enable all transmission interrupts and clear current status */
-    _i2c_hw[dev].regs->int_ena.val &= ~transfer_int_mask;
-    _i2c_hw[dev].regs->int_ena.val |= transfer_int_mask;
-    _i2c_hw[dev].regs->int_clr.val  = transfer_int_mask;
+    i2c_hal_clr_intsts_mask(&_i2c_hw[dev], I2C_LL_MASTER_INT);
+    i2c_hal_enable_intr_mask(&_i2c_hw[dev], I2C_LL_MASTER_INT);
 
     /* set a timer for the case the I2C hardware gets stuck */
-    ztimer_t i2c_timeout = {};
-    i2c_timeout.callback = _i2c_transfer_timeout;
-    i2c_timeout.arg = (void*)(uintptr_t)dev;
-    ztimer_set(ZTIMER_MSEC, &i2c_timeout, I2C_TRANSFER_TIMEOUT_MS);
+#if defined(MODULE_ZTIMER_MSEC)
+    uint32_t timeout = ((I2C_TRANSFER_TIMEOUT * KHZ(1)) / _i2c_bus[dev].clk_freq) + 1;
+    ztimer_t timer = { .callback = _i2c_transfer_timeout,
+                       .arg = (void*)(uintptr_t)dev };
+    ztimer_set(ZTIMER_MSEC, &timer, timeout);
+#endif
+
+    /* start the execution of commands in command pipeline */
+    _i2c_bus[dev].status = 0;
 
-    /* start execution of commands in command pipeline registers */
-    _i2c_bus[dev].pid = thread_getpid();
-    _i2c_bus[dev].results = 0;
-    _i2c_hw[dev].regs->ctr.trans_start = 0;
-    _i2c_hw[dev].regs->ctr.trans_start = 1;
+    i2c_hal_update_config(&_i2c_hw[dev]);
+    i2c_hal_trans_start(&_i2c_hw[dev]);
 
     /* wait for transfer results and remove timeout timer*/
-    thread_flags_wait_one(I2C_THREAD_FLAG);
-    ztimer_remove(ZTIMER_MSEC, &i2c_timeout);
+    mutex_lock(&_i2c_bus[dev].cmd_lock);
 
-    /* returned from transmission */
-    DEBUG("%s results=%08x\n", __func__, _i2c_bus[dev].results);
+    critical_enter();
+#if defined(MODULE_ZTIMER_MSEC)
+    ztimer_remove(ZTIMER_MSEC, &timer);
+#endif
+    critical_exit();
```

Comment:
Does `ztimer_remove()` really need this kind of protection? IMHO it would be a bug in ztimer if the API can only be called with interrupts disabled.

Reply:
Ok, `ztimer_remove` disables and enables the interrupts. I removed it.